### ISSUE TITLE
DSL.and(noCondition()) and DSL.or(noCondition()) generates a "1 = 1" or "1 = 0" predicate

### ISSUE THREAD

**Issue by vkolotov** [jOOQ/jOOQ#9998]
### Expected behavior and actual behavior:

Sometimes DSL.noCondition() is making some queries x5 slower than if there was not any condition.

Expected behaviour - noCondition() should not add any condition, e.g. empty string. If no other condition present, then "where" clause should not be rendered.

### Steps to reproduce the problem: 

```java
select(...).from(...).where(DSL.noCondition())
```

This statement ^^ generates a sql query like that:

```sql
select <columns> from <table> where 1=1
```

Adding 1=1 where clause, in some cases, drastically slows down query execution, in my case it was 1.5s vs 10s.

See more details [here](https://dba.stackexchange.com/questions/33937/good-bad-or-indifferent-where-1-1)

### Versions:

- jOOQ: 3.12.3
- Java: 11
- Database (include vendor): MS SQL 12
- OS: ubuntu
- JDBC Driver (include name if inofficial driver): mssql-jdbc-7.2.1.jre11.jar

**Comment by lukaseder** [comment]
Thank you very much for your report. I've noticed this as well in SQL Server, in the past. Quite regrettable: https://blog.jooq.org/2017/09/28/10-cool-sql-optimisations-that-do-not-depend-on-the-cost-model/#top4

But are you sure you're using `noCondition()` and not `trueCondition()`? `noCondition()` should not generate anything at all, whereas `trueCondition()` is used when this outcome is desired.

**Comment by lukaseder** [comment]
It might be a good idea to hint at this possibility in the Javadoc of `trueCondition()`

**Comment by vkolotov** [comment]
Yes, I'm sure I'm using `noCondition()`. Here is what making 1=1 in there: 
https://github.com/jOOQ/jOOQ/blob/5108d7c1925043a172f2ee66d00490590cca280d/jOOQ/src/main/java/org/jooq/impl/NoCondition.java#L58

**Comment by lukaseder** [comment]
That's what it renders when it gets forced to render itself, but it shouldn't do it, if it's not necessary. I'll try to reproduce it with jOOQ 3.12

**Comment by vkolotov** [comment]
You may need to add another condition to reproduce it though, like:

```
select(...).from(...)
.where(
    and(
        field.eq("test"), 
        noCondition()
    )
)
```

**Comment by lukaseder** [comment]
This doesn't reproduce it:

```java
SQLDialect family = SQLDialect.SQLSERVER;
DSLContext ctx = DSL.using(family);
System.out.println(ctx.selectOne().from("t").where(DSL.noCondition()));
```

**Comment by vkolotov** [comment]
I'm using the opensource jooq, default dialect.

**Comment by lukaseder** [comment]
This also doesn't reproduce it:

```java
SQLDialect family = SQLDialect.SQLSERVER;
DSLContext ctx = DSL.using(family);
System.out.println(ctx.selectOne().from("t").where(and(field("x").eq("test"), noCondition())));
```

(I'm using 3.12.4)

**Comment by lukaseder** [comment]
This also doesn't produce it:

```java
SQLDialect family = SQLDialect.DEFAULT;
DSLContext ctx = DSL.using(family);
System.out.println(ctx.selectOne().from("t").where(and(field("x").eq("test"), noCondition())));
```

I'm getting:

```sql
select 1 "one"
from t
where x = 'test'
```

However, why do you use the default dialect on SQL Server? That probably doesn't work very well...

**Comment by lukaseder** [comment]
No difference between the jOOQ Open Source Edition and the commercial distributions either...

**Comment by lukaseder** [comment]
Perhaps an MCVE would help reproduce this problem? https://github.com/jOOQ/jOOQ-mcve

**Comment by vkolotov** [comment]
This 
```
jooq.select(DSL.asterisk())
        .from("table")
        .where(and(and(noCondition()))).getSQL()
```
Renders
```
select * from table where 1 = 1
```

**Comment by lukaseder** [comment]
Ah yes, I see. Thanks for your help.

**Comment by vkolotov** [comment]
I'm using jooq as a query builder exclusively.
Thank you very much for your help.

Looks like I'll need to avoid using nested and's for now.

**Comment by lukaseder** [comment]
This seems to be a valid workaround: `and(noCondition(), noCondition())`

**Comment by vkolotov** [comment]
Ah yeah! That's better! Thanks!

**Comment by lukaseder** [comment]
Ah, that workaround only works because it's a different overload. The overloads taking arrays or collections don't profit from this workaround

**Comment by vkolotov** [comment]
Confirming that workaround does not work on vararg methods, e.g. and(Condition... parts)

**Comment by lukaseder** [comment]
But if you use `Condition.and()` instead of `DSL.and()`, then you'd profit from the workaround. Perhaps, you could write a `Stream` reduction using `Condition::and` with the `DSL.noCondition()` as the identity?

**Comment by vkolotov** [comment]
Yep, this is what I did, a stream of Options with reduction to DSL.and or else `noCondition()`

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.0. Backport scheduled for 3.13.2 (#10002) and 3.12.5 (#10003). Thanks again for your report.
